File: src/cppinterface/cb_boxmodel.jl
@doc raw"""
    cb_clear!(self::CBBoxModel)

resets all data to initial status of this class, also the bundle parameters
"""
cb_clear!(self::CBBoxModel) = @ccall libcb.cb_boxmodel_clear(self.data::Ptr{Cvoid})::Cvoid

@doc raw"""
    CBBoxModel(fo::Union{<:CBBoxOracle,Nothing}, fun_factor::Real = 1., fun_task::CBFunctionTask = cbft_objective_function, cbinc::Integer = -1)

construct a model for the MatrixBoxOracle pointed to by @a fo
"""
CBBoxModel(fo::Union{<:CBBoxOracle,Nothing}, fun_factor::Real = 1., fun_task::CBFunctionTask = cbft_objective_function, cbinc::Integer = -1) = CBBoxModel(@ccall libcb.cb_boxmodel_new((isnothing(fo) ? C_NULL : fo.data)::Ptr{Cvoid}, fun_factor::Cdouble, fun_task::Cint, cbinc::Cint)::Ptr{Cvoid})

@doc raw"""
    cb_get_oracle_object!(self::CBBoxModel)

returns the oracle
"""
cb_get_oracle_object!(self::CBBoxModel) = CBModifiableOracleObject(@ccall libcb.cb_boxmodel_get_oracle_object(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_lb_function!(self::CBBoxModel, y_id::Integer, y::CBMatrix)

see SumBlockModel::lb_function
"""
cb_lb_function!(self::CBBoxModel, y_id::Integer, y::CBMatrix) = @ccall libcb.cb_boxmodel_lb_function(self.data::Ptr{Cvoid}, y_id::Cint, y.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_data!(self::CBBoxModel)

see SumBlockModel::get_data
"""
cb_get_data!(self::CBBoxModel) = CBBundleData(@ccall libcb.cb_boxmodel_get_data(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_data(self::CBBoxModel)

see SumBlockModel::get_data
"""
cb_get_data(self::CBBoxModel) = CBBundleData(@ccall libcb.cb_boxmodel_get_data2(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_set_data!(self::CBBoxModel, bd::Union{<:CBBundleData,Nothing})

see SumBlockModel::set_data
"""
cb_set_data!(self::CBBoxModel, bd::Union{<:CBBundleData,Nothing}) = @ccall libcb.cb_boxmodel_set_data(self.data::Ptr{Cvoid}, (isnothing(bd) ? C_NULL : bd.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_approximate_primal(self::CBBoxModel)

see SumBlockModel::get_approximate_primal
"""
cb_get_approximate_primal(self::CBBoxModel) = CBPrimalData(@ccall libcb.cb_boxmodel_get_approximate_primal(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_center_primal(self::CBBoxModel)

see SumBlockModel::get_cneter_primal
"""
cb_get_center_primal(self::CBBoxModel) = CBPrimalData(@ccall libcb.cb_boxmodel_get_center_primal(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_candidate_primal(self::CBBoxModel)

see SumBlockModel::get_candidate_primal
"""
cb_get_candidate_primal(self::CBBoxModel) = CBPrimalData(@ccall libcb.cb_boxmodel_get_candidate_primal(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_call_primal_extender!(self::CBBoxModel, prex::CBPrimalExtender)

see SumBlockModel::call_primal_extender
"""
cb_call_primal_extender!(self::CBBoxModel, prex::CBPrimalExtender) = @ccall libcb.cb_boxmodel_call_primal_extender(self.data::Ptr{Cvoid}, prex.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_set_bundle_parameters!(self::CBBoxModel, bp::CBBundleParameters)

see SumBlockModel::set_bundle_parameters
"""
cb_set_bundle_parameters!(self::CBBoxModel, bp::CBBundleParameters) = @ccall libcb.cb_boxmodel_set_bundle_parameters(self.data::Ptr{Cvoid}, bp.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_bundle_parameters(self::CBBoxModel)

see SumBlockModel::get_bundle_parameters
"""
cb_get_bundle_parameters(self::CBBoxModel) = CBBundleParameters(@ccall libcb.cb_boxmodel_get_bundle_parameters(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_ret_code(self::CBBoxModel)

see SumBlockModel::get_ret_code()
"""
cb_get_ret_code(self::CBBoxModel) = @ccall libcb.cb_boxmodel_get_ret_code(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_set_out!(self::CBBoxModel, pril::Integer = 1)

set output and outputlevel of warnings and errors recursively, see CBout
"""
cb_set_out!(self::CBBoxModel, pril::Integer = 1) = @ccall libcb.cb_boxmodel_set_out(self.data::Ptr{Cvoid}, pril::Cint)::Cvoid

File: ConicBundle/cppinterface/cb_boxmodel.cpp
dll void cb_boxmodel_destroy(BoxModel* self) {
  delete self;
}

dll void cb_boxmodel_clear(BoxModel* self) {
  self->clear();
}

dll BoxModel* cb_boxmodel_new(BoxOracle* fo, Real fun_factor = 1., int fun_task = ObjectiveFunction, int cbinc = -1) {
  return new BoxModel(fo, fun_factor, (FunctionTask)fun_task, 0, cbinc);
}

dll ModifiableOracleObject* cb_boxmodel_get_oracle_object(BoxModel* self) {
  return self->get_oracle_object();
}

dll Real cb_boxmodel_lb_function(BoxModel* self, Integer y_id, const Matrix* y) {
  return self->lb_function(y_id, *y);
}

dll BundleData* cb_boxmodel_get_data(BoxModel* self) {
  return self->get_data();
}

dll const BundleData* cb_boxmodel_get_data2(const BoxModel* self) {
  return self->get_data();
}

dll int cb_boxmodel_set_data(BoxModel* self, BundleData* bd) {
  return self->set_data(bd);
}

dll const PrimalData* cb_boxmodel_get_approximate_primal(const BoxModel* self) {
  return self->get_approximate_primal();
}

dll const PrimalData* cb_boxmodel_get_center_primal(const BoxModel* self) {
  return self->get_center_primal();
}

dll const PrimalData* cb_boxmodel_get_candidate_primal(const BoxModel* self) {
  return self->get_candidate_primal();
}

dll int cb_boxmodel_call_primal_extender(BoxModel* self, PrimalExtender* prex) {
  return self->call_primal_extender(*prex);
}

dll int cb_boxmodel_set_bundle_parameters(BoxModel* self, const BundleParameters* bp) {
  return self->set_bundle_parameters(*bp);
}

dll BundleParameters* cb_boxmodel_get_bundle_parameters(const BoxModel* self) {
  return self->get_bundle_parameters();
}

dll int cb_boxmodel_get_ret_code(const BoxModel* self) {
  return self->get_ret_code();
}

dll void cb_boxmodel_set_out(BoxModel* self, int pril = 1) {
  self->set_out(&std::cout, pril);
}

